// @szhsin/react-menu/dist/index.js
function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));
var ReactDOM = _interopDefault(require('react-dom'));

var menuContainerClass = 'rc-menu-container';
var menuClass = 'rc-menu';
var menuButtonClass = 'rc-menu-button';
var menuArrowClass = 'arrow';
var menuItemClass = 'item';
var menuDividerClass = 'divider';
var menuHeaderClass = 'header';
var subMenuClass = 'submenu';
var radioGroupClass = 'radio-group';
var initialHoverIndex = -1;
var MenuListItemContext = /*#__PURE__*/React__default.createContext({});
var MenuListContext = /*#__PURE__*/React__default.createContext();
var EventHandlersContext = /*#__PURE__*/React__default.createContext({});
var RadioGroupContext = /*#__PURE__*/React__default.createContext({});
var SettingsContext = /*#__PURE__*/React__default.createContext({});
var SUBMENU_CLOSE_DELAY = 160;
var SUBMENU_OPEN_DELAY = 160;
var Keys = Object.freeze({
  'ENTER': 'Enter',
  'ESC': 'Escape',
  'SPACE': ' ',
  'HOME': 'Home',
  'END': 'End',
  'LEFT': 'ArrowLeft',
  'RIGHT': 'ArrowRight',
  'UP': 'ArrowUp',
  'DOWN': 'ArrowDown'
});
var HoverIndexActionTypes = Object.freeze({
  'RESET': 'HOVER_INDEX_RESET',
  'SET': 'HOVER_INDEX_SET',
  'UNSET': 'HOVER_INDEX_UNSET',
  'INCREASE': 'HOVER_INDEX_INCREASE',
  'DECREASE': 'HOVER_INDEX_DECREASE',
  'FIRST': 'HOVER_INDEX_FIRST',
  'LAST': 'HOVER_INDEX_LAST'
});
var SubmenuActionTypes = Object.freeze({
  'OPEN': 'SUBMENU_OPEN',
  'CLOSE': 'SUBMENU_CLOSE'
});
var CloseReason = Object.freeze({
  'CLICK': 'click',
  'CANCEL': 'cancel',
  'BLUR': 'blur',
  'SCROLL': 'scroll'
});

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

var stylePropTypes = function stylePropTypes(name) {
  var _ref;

  return _ref = {}, _ref[name ? name + "ClassName" : 'className'] = PropTypes.oneOfType([PropTypes.string, PropTypes.func]), _ref[name ? name + "Styles" : 'styles'] = PropTypes.oneOfType([PropTypes.object, PropTypes.func]), _ref;
};
var sharedMenuPropTypes = _extends({}, stylePropTypes(), stylePropTypes('arrow'), {
  'aria-label': PropTypes.string,
  arrow: PropTypes.bool,
  offsetX: PropTypes.number,
  offsetY: PropTypes.number,
  align: PropTypes.oneOf(['start', 'center', 'end']),
  direction: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
  position: PropTypes.oneOf(['auto', 'anchor', 'initial']),
  overflow: PropTypes.oneOf(['auto', 'visible', 'hidden']),
  children: PropTypes.node.isRequired
});
var menuPropTypesBase = _extends({}, sharedMenuPropTypes, {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  animation: PropTypes.bool,
  debugging: PropTypes.bool,
  boundingBoxRef: PropTypes.object,
  boundingBoxPadding: PropTypes.string,
  reposition: PropTypes.oneOf(['auto', 'initial']),
  repositionFlag: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  viewScroll: PropTypes.oneOf(['auto', 'close', 'initial']),
  portal: PropTypes.bool,
  theming: PropTypes.string,
  onClick: PropTypes.func
});
var sharedMenuDefaultProp = {
  offsetX: 0,
  offsetY: 0,
  align: 'start',
  direction: 'bottom',
  position: 'auto',
  overflow: 'visible'
};
var menuDefaultPropsBase = _extends({}, sharedMenuDefaultProp, {
  animation: true,
  reposition: 'auto',
  viewScroll: 'initial'
});

var defineName = function defineName(component, name) {
  return name ? Object.defineProperty(component, '_reactMenu', {
    value: name,
    writable: false
  }) : component;
};
var getName = function getName(component) {
  return component && component['_reactMenu'];
};
var applyHOC = function applyHOC(HOC) {
  return function () {
    return defineName(HOC.apply(void 0, arguments), getName(arguments.length <= 0 ? undefined : arguments[0]));
  };
};
var applyStatics = function applyStatics(sourceComponent) {
  return function (wrappedComponent) {
    return defineName(wrappedComponent, getName(sourceComponent));
  };
};
var safeCall = function safeCall(fn) {
  for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  return typeof fn === 'function' ? fn.apply(void 0, args) : fn;
};
var attachHandlerProps = function attachHandlerProps(handlers, props) {
  var result = {};

  var _loop = function _loop() {
    var handlerName = _Object$keys[_i];
    var handler = handlers[handlerName];
    var propHandler = props[handlerName];
    var attachedHandler = void 0;

    if (typeof propHandler === 'function') {
      attachedHandler = function attachedHandler(e) {
        propHandler(e);
        handler(e);
      };
    } else {
      attachedHandler = handler;
    }

    result[handlerName] = attachedHandler;
  };

  for (var _i = 0, _Object$keys = Object.keys(handlers); _i < _Object$keys.length; _i++) {
    _loop();
  }

  return result;
};
var parsePadding = function parsePadding(paddingStr) {
  if (typeof paddingStr !== 'string') return {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  };
  var padding = paddingStr.trim().split(/\s+/, 4).map(parseFloat);
  var top = !isNaN(padding[0]) ? padding[0] : 0;
  var right = !isNaN(padding[1]) ? padding[1] : top;
  return {
    top: top,
    right: right,
    bottom: !isNaN(padding[2]) ? padding[2] : top,
    left: !isNaN(padding[3]) ? padding[3] : right
  };
};
var bem = function bem(block, element, modifiers) {
  if (modifiers === void 0) {
    modifiers = {};
  }

  return function (userClassName, userModifiers) {
    var blockElement = element ? block + "__" + element : block;
    var className = blockElement;

    for (var _i2 = 0, _Object$keys2 = Object.keys(modifiers); _i2 < _Object$keys2.length; _i2++) {
      var name = _Object$keys2[_i2];
      var value = modifiers[name];

      if (value) {
        className += " " + blockElement + "--";
        className += value === true ? name : name + "-" + value;
      }
    }

    if (typeof userClassName === 'function') {
      userClassName = userClassName(userModifiers || modifiers);
    }

    if (typeof userClassName === 'string') {
      userClassName = userClassName.trim();
      if (userClassName) className += " " + userClassName;
    }

    return className;
  };
};

var isObject = function isObject(obj) {
  return obj && typeof obj === 'object';
};

var sanitiseKey = function sanitiseKey(key) {
  return key.charAt(0) === '$' ? key.slice(1) : key;
};

var flatStyles = function flatStyles(styles, modifiers) {
  if (typeof styles === 'function') return styles(modifiers);
  if (!isObject(styles)) return undefined;
  if (!modifiers) return styles;
  var style = {};

  for (var _i3 = 0, _Object$keys3 = Object.keys(styles); _i3 < _Object$keys3.length; _i3++) {
    var prop = _Object$keys3[_i3];
    var value = styles[prop];

    if (isObject(value)) {
      var modifierValue = modifiers[sanitiseKey(prop)];

      if (typeof modifierValue === 'string') {
        for (var _i4 = 0, _Object$keys4 = Object.keys(value); _i4 < _Object$keys4.length; _i4++) {
          var nestedProp = _Object$keys4[_i4];
          var nestedValue = value[nestedProp];

          if (isObject(nestedValue)) {
            if (sanitiseKey(nestedProp) === modifierValue) {
              Object.assign(style, nestedValue);
            }
          } else {
            style[nestedProp] = nestedValue;
          }
        }
      } else if (modifierValue) {
        Object.assign(style, value);
      }
    } else {
      style[prop] = value;
    }
  }

  return style;
};
var getScrollAncestor = function getScrollAncestor(node) {
  while (node && node !== document.body) {
    var _getComputedStyle = getComputedStyle(node),
        overflow = _getComputedStyle.overflow,
        overflowX = _getComputedStyle.overflowX,
        overflowY = _getComputedStyle.overflowY;

    if (/auto|scroll|overlay|hidden/.test(overflow + overflowY + overflowX)) return node;
    node = node.parentNode;
  }

  return window;
};
var floatEqual = function floatEqual(a, b) {
  return Math.abs(a - b) < 0.001;
};
var isProd = process && process.env && process.env.NODE_ENV === 'production';

var useActiveState = function useActiveState(isHovering, isDisabled, customKey) {
  var _useState = React.useState(false),
      active = _useState[0],
      setActive = _useState[1];

  var activeKeys = React.useMemo(function () {
    return [Keys.SPACE, Keys.ENTER].concat(customKey ? [customKey] : []);
  }, [customKey]);
  var cancelActive = React.useCallback(function () {
    return setActive(false);
  }, []);
  return {
    isActive: active,
    onPointerDown: React.useCallback(function () {
      if (!isDisabled) setActive(true);
    }, [isDisabled]),
    onPointerUp: cancelActive,
    onPointerLeave: cancelActive,
    onKeyDown: React.useCallback(function (e) {
      if (isHovering && !isDisabled && activeKeys.includes(e.key)) {
        setActive(true);
      }
    }, [isHovering, isDisabled, activeKeys]),
    onKeyUp: React.useCallback(function (e) {
      if (activeKeys.includes(e.key)) {
        setActive(false);
      }
    }, [activeKeys]),
    onBlur: React.useCallback(function (e) {
      if (!e.currentTarget.contains(e.relatedTarget)) {
        setActive(false);
      }
    }, [])
  };
};

var useItemState = function useItemState(isDisabled, index) {
  var ref = React.useRef(null);

  var _useContext = React.useContext(MenuListItemContext),
      isParentOpen = _useContext.isParentOpen,
      hoverIndex = _useContext.hoverIndex,
      isSubmenuOpen = _useContext.isSubmenuOpen,
      dispatch = _useContext.dispatch;

  var isHovering = hoverIndex === index;
  var timeoutId = React.useRef();
  React.useEffect(function () {
    if (isHovering && isParentOpen) {
      ref.current && ref.current.focus();
    }
  }, [isHovering, isParentOpen]);
  var setHover = React.useCallback(function () {
    if (!isDisabled) dispatch({
      type: HoverIndexActionTypes.SET,
      index: index
    });
  }, [isDisabled, dispatch, index]);
  var onBlur = React.useCallback(function (e) {
    if (!e.currentTarget.contains(e.relatedTarget)) {
      dispatch({
        type: HoverIndexActionTypes.UNSET,
        index: index
      });
    }
  }, [dispatch, index]);
  var onMouseEnter = React.useCallback(function () {
    if (isSubmenuOpen) {
      timeoutId.current = setTimeout(setHover, SUBMENU_CLOSE_DELAY);
    } else {
      setHover();
    }
  }, [isSubmenuOpen, setHover]);
  var onMouseLeave = React.useCallback(function (_, keepHover) {
    timeoutId.current && clearTimeout(timeoutId.current);
    if (!keepHover) dispatch({
      type: HoverIndexActionTypes.UNSET,
      index: index
    });
  }, [dispatch, index]);
  return {
    ref: ref,
    isHovering: isHovering,
    setHover: setHover,
    onBlur: onBlur,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave
  };
};

var useMenuChange = function useMenuChange(onChange, isOpen) {
  var prevOpen = React.useRef(isOpen);
  React.useEffect(function () {
    if (prevOpen.current !== isOpen) safeCall(onChange, {
      open: isOpen
    });
    prevOpen.current = isOpen;
  }, [onChange, isOpen]);
};

var FocusPositions = Object.freeze({
  'INITIAL': 'initial',
  'FIRST': 'first',
  'LAST': 'last'
});
var MenuStates = Object.freeze({
  'UNMOUNTED': 0,
  'CLOSED': 1,
  'OPEN': 2
});
var MenuStateActionType = Object.freeze({
  'OPEN': 0,
  'CLOSE': 1,
  'TOGGLE': 2
});
var useMenuState = function useMenuState(keepMounted) {
  if (keepMounted === void 0) {
    keepMounted = true;
  }

  var _useState = React.useState({
    position: FocusPositions.INITIAL
  }),
      menuItemFocus = _useState[0],
      setMenuItemFocus = _useState[1];

  var _useReducer = React.useReducer(menuStateReducer, MenuStates.UNMOUNTED),
      menuState = _useReducer[0],
      dispatch = _useReducer[1];

  function menuStateReducer(state, _ref) {
    var type = _ref.type;

    switch (type) {
      case MenuStateActionType.OPEN:
        return MenuStates.OPEN;

      case MenuStateActionType.CLOSE:
        if (state === MenuStates.OPEN) {
          return keepMounted ? MenuStates.CLOSED : MenuStates.UNMOUNTED;
        } else {
          return state;
        }

      case MenuStateActionType.TOGGLE:
        if (state === MenuStates.OPEN) {
          return keepMounted ? MenuStates.CLOSED : MenuStates.UNMOUNTED;
        } else {
          return MenuStates.OPEN;
        }

      default:
        throw new Error('menuStateReducer: unknown action type');
    }
  }

  return {
    isMounted: menuState !== MenuStates.UNMOUNTED,
    isOpen: menuState === MenuStates.OPEN,
    menuItemFocus: menuItemFocus,
    openMenu: React.useCallback(function (menuItemFocus) {
      if (menuItemFocus === void 0) {
        menuItemFocus = FocusPositions.INITIAL;
      }

      setMenuItemFocus({
        position: menuItemFocus
      });
      dispatch({
        type: MenuStateActionType.OPEN
      });
    }, []),
    closeMenu: React.useCallback(function () {
      return dispatch({
        type: MenuStateActionType.CLOSE
      });
    }, []),
    toggleMenu: React.useCallback(function (menuItemFocus) {
      if (menuItemFocus === void 0) {
        menuItemFocus = FocusPositions.INITIAL;
      }

      setMenuItemFocus({
        position: menuItemFocus
      });
      dispatch({
        type: MenuStateActionType.TOGGLE
      });
    }, [])
  };
};

var useIsomorphicLayoutEffect = typeof window !== 'undefined' && typeof window.document !== 'undefined' && typeof window.document.createElement !== 'undefined' ? React.useLayoutEffect : React.useEffect;

var MenuButton = defineName( /*#__PURE__*/React__default.memo( /*#__PURE__*/React__default.forwardRef(function MenuButton(_ref, ref) {
  var className = _ref.className,
      styles = _ref.styles,
      isOpen = _ref.isOpen,
      disabled = _ref.disabled,
      children = _ref.children,
      restProps = _objectWithoutPropertiesLoose(_ref, ["className", "styles", "isOpen", "disabled", "children"]);

  var modifiers = Object.freeze({
    open: isOpen
  });
  return /*#__PURE__*/React__default.createElement("button", _extends({
    "aria-haspopup": "true",
    "aria-expanded": isOpen,
    "aria-disabled": disabled || undefined,
    disabled: disabled
  }, restProps, {
    ref: ref,
    className: bem(menuButtonClass, null, modifiers)(className),
    style: flatStyles(styles, modifiers)
  }), children);
})), 'MenuButton');
MenuButton.propTypes = _extends({}, stylePropTypes(), {
  isOpen: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func
});

var MenuList = defineName( /*#__PURE__*/React__default.memo(function MenuList(_ref) {
  var ariaLabel = _ref.ariaLabel,
      className = _ref.className,
      styles = _ref.styles,
      arrowClassName = _ref.arrowClassName,
      arrowStyles = _ref.arrowStyles,
      anchorPoint = _ref.anchorPoint,
      anchorRef = _ref.anchorRef,
      containerRef = _ref.containerRef,
      arrow = _ref.arrow,
      align = _ref.align,
      direction = _ref.direction,
      position = _ref.position,
      overflow = _ref.overflow,
      repositionFlag = _ref.repositionFlag,
      _ref$captureFocus = _ref.captureFocus,
      captureFocus = _ref$captureFocus === void 0 ? true : _ref$captureFocus,
      isOpen = _ref.isOpen,
      isDisabled = _ref.isDisabled,
      menuItemFocus = _ref.menuItemFocus,
      offsetX = _ref.offsetX,
      offsetY = _ref.offsetY,
      children = _ref.children,
      onClose = _ref.onClose,
      restProps = _objectWithoutPropertiesLoose(_ref, ["ariaLabel", "className", "styles", "arrowClassName", "arrowStyles", "anchorPoint", "anchorRef", "containerRef", "arrow", "align", "direction", "position", "overflow", "repositionFlag", "captureFocus", "isOpen", "isDisabled", "menuItemFocus", "offsetX", "offsetY", "children", "onClose"]);

  var _useState = React.useState({
    x: 0,
    y: 0
  }),
      menuPosition = _useState[0],
      setMenuPosition = _useState[1];

  var _useState2 = React.useState({}),
      arrowPosition = _useState2[0],
      setArrowPosition = _useState2[1];

  var _useState3 = React.useState(-1),
      maxHeight = _useState3[0],
      setMaxHeight = _useState3[1];

  var _useState4 = React.useState(false),
      isClosing = _useState4[0],
      setClosing = _useState4[1];

  var _useState5 = React.useState(direction),
      expandedDirection = _useState5[0],
      setExpandedDirection = _useState5[1];

  var _useContext = React.useContext(SettingsContext),
      animation = _useContext.animation,
      boundingBoxRef = _useContext.boundingBoxRef,
      boundingBoxPadding = _useContext.boundingBoxPadding,
      menuRootRef = _useContext.menuRootRef,
      scrollingRef = _useContext.scrollingRef,
      reposition = _useContext.reposition,
      viewScroll = _useContext.viewScroll;

  var menuRef = React.useRef(null);
  var arrowRef = React.useRef(null);
  var menuItemsCount = React.useRef(0);
  var latestOpen = React.useRef(isOpen);
  var latestMenuSize = React.useRef({
    width: 0,
    height: 0
  });
  var latestHandlePosition = React.useRef(function () {});
  var reposFlag = React.useContext(MenuListContext) || repositionFlag;

  var _useReducer = React.useReducer(function (c) {
    return c + 1;
  }, 1),
      reposSubmenu = _useReducer[0],
      forceReposSubmenu = _useReducer[1];

  var _useReducer2 = React.useReducer(reducer, {
    hoverIndex: initialHoverIndex,
    openSubmenuCount: 0
  }),
      _useReducer2$ = _useReducer2[0],
      hoverIndex = _useReducer2$.hoverIndex,
      openSubmenuCount = _useReducer2$.openSubmenuCount,
      dispatch = _useReducer2[1];

  function reducer(_ref2, action) {
    var hoverIndex = _ref2.hoverIndex,
        openSubmenuCount = _ref2.openSubmenuCount;
    return {
      hoverIndex: hoverIndexReducer(hoverIndex, action),
      openSubmenuCount: submenuCountReducer(openSubmenuCount, action)
    };
  }

  function hoverIndexReducer(state, _ref3) {
    var type = _ref3.type,
        index = _ref3.index;

    switch (type) {
      case HoverIndexActionTypes.RESET:
        return initialHoverIndex;

      case HoverIndexActionTypes.SET:
        return index;

      case HoverIndexActionTypes.UNSET:
        return state === index ? initialHoverIndex : state;

      case HoverIndexActionTypes.DECREASE:
        {
          var i = state;
          i--;
          if (i < 0) i = menuItemsCount.current - 1;
          return i;
        }

      case HoverIndexActionTypes.INCREASE:
        {
          var _i = state;
          _i++;
          if (_i >= menuItemsCount.current) _i = 0;
          return _i;
        }

      case HoverIndexActionTypes.FIRST:
        return menuItemsCount.current > 0 ? 0 : initialHoverIndex;

      case HoverIndexActionTypes.LAST:
        return menuItemsCount.current > 0 ? menuItemsCount.current - 1 : initialHoverIndex;

      default:
        return state;
    }
  }

  var menuItems = React.useMemo(function () {
    var index = 0;
    var permittedChildren = ['MenuDivider', 'MenuHeader', 'MenuItem', 'FocusableItem', 'MenuRadioGroup', 'SubMenu'];

    var validateChildren = function validateChildren(parent, child, permitted) {
      if (!child) return false;

      if (!permitted.includes(getName(child.type))) {
        console.warn((child.type || child) + " is ignored.\n", "The permitted children inside a " + parent + " are " + permitted.join(', ') + ".", 'If you create HOC of these components, you can use the applyHOC or applyStatics helper, see more at: https://szhsin.github.io/react-menu/docs#utils-apply-hoc');
        return false;
      }

      return true;
    };

    var items = React__default.Children.map(children, function (child) {
      if (!validateChildren('Menu or SubMenu', child, permittedChildren)) return null;
      var componentName = getName(child.type);

      if (componentName === 'MenuDivider' || componentName === 'MenuHeader') {
        return child;
      } else if (componentName === 'MenuRadioGroup') {
        var _permittedChildren = ['MenuItem'];
        var props = {
          type: 'radio'
        };
        var radioItems = React__default.Children.map(child.props.children, function (radioChild) {
          if (!validateChildren('MenuRadioGroup', radioChild, _permittedChildren)) return null;
          return radioChild.props.disabled ? /*#__PURE__*/React__default.cloneElement(radioChild, props) : /*#__PURE__*/React__default.cloneElement(radioChild, _extends({}, props, {
            index: index++
          }));
        });
        return /*#__PURE__*/React__default.cloneElement(child, {
          children: radioItems
        });
      } else {
        return child.props.disabled ? child : /*#__PURE__*/React__default.cloneElement(child, {
          index: index++
        });
      }
    });
    menuItemsCount.current = index;
    return items;
  }, [children]);

  var handleKeyDown = function handleKeyDown(e) {
    var handled = false;

    switch (e.key) {
      case Keys.HOME:
        dispatch({
          type: HoverIndexActionTypes.FIRST
        });
        handled = true;
        break;

      case Keys.END:
        dispatch({
          type: HoverIndexActionTypes.LAST
        });
        handled = true;
        break;

      case Keys.UP:
        dispatch({
          type: HoverIndexActionTypes.DECREASE
        });
        handled = true;
        break;

      case Keys.DOWN:
        dispatch({
          type: HoverIndexActionTypes.INCREASE
        });
        handled = true;
        break;

      case Keys.SPACE:
        if (e.target && e.target.className.includes(menuClass)) {
          e.preventDefault();
        }

        break;
    }

    if (handled) {
      e.preventDefault();
      e.stopPropagation();
    }
  };

  var handleAnimationEnd = function handleAnimationEnd() {
    if (isClosing) {
      setClosing(false);
      setMaxHeight(-1);
    }
  };

  var positionHelpers = React.useCallback(function (boundingBoxRef, boundingBoxPadding) {
    var menuRect = menuRef.current.getBoundingClientRect();
    var containerRect = containerRef.current.getBoundingClientRect();

    if (!scrollingRef.current) {
      scrollingRef.current = boundingBoxRef ? boundingBoxRef.current : getScrollAncestor(menuRootRef.current);
    }

    var boundingRect = scrollingRef.current === window ? {
      left: 0,
      top: 0,
      right: document.documentElement.clientWidth,
      bottom: window.innerHeight
    } : scrollingRef.current.getBoundingClientRect();
    var padding = parsePadding(boundingBoxPadding);

    var getLeftOverflow = function getLeftOverflow(x) {
      return x + containerRect.left - boundingRect.left - padding.left;
    };

    var getRightOverflow = function getRightOverflow(x) {
      return x + containerRect.left + menuRect.width - boundingRect.right + padding.right;
    };

    var getTopOverflow = function getTopOverflow(y) {
      return y + containerRect.top - boundingRect.top - padding.top;
    };

    var getBottomOverflow = function getBottomOverflow(y) {
      return y + containerRect.top + menuRect.height - boundingRect.bottom + padding.bottom;
    };

    var confineHorizontally = function confineHorizontally(x) {
      var leftOverflow = getLeftOverflow(x);

      if (leftOverflow < 0) {
        x -= leftOverflow;
      } else {
        var rightOverflow = getRightOverflow(x);

        if (rightOverflow > 0) {
          x -= rightOverflow;
          leftOverflow = getLeftOverflow(x);
          if (leftOverflow < 0) x -= leftOverflow;
        }
      }

      return x;
    };

    var confineVertically = function confineVertically(y) {
      var topOverflow = getTopOverflow(y);

      if (topOverflow < 0) {
        y -= topOverflow;
      } else {
        var bottomOverflow = getBottomOverflow(y);

        if (bottomOverflow > 0) {
          y -= bottomOverflow;
          topOverflow = getTopOverflow(y);
          if (topOverflow < 0) y -= topOverflow;
        }
      }

      return y;
    };

    return {
      menuRect: menuRect,
      containerRect: containerRect,
      getLeftOverflow: getLeftOverflow,
      getRightOverflow: getRightOverflow,
      getTopOverflow: getTopOverflow,
      getBottomOverflow: getBottomOverflow,
      confineHorizontally: confineHorizontally,
      confineVertically: confineVertically
    };
  }, [containerRef, menuRootRef, scrollingRef]);
  var placeArrowX = React.useCallback(function (menuX, anchorRect, containerRect, menuRect) {
    if (!arrow) return;
    var x = anchorRect.left - containerRect.left - menuX + anchorRect.width / 2;
    var offset = arrowRef.current.offsetWidth * 1.25;
    x = Math.max(offset, x);
    x = Math.min(x, menuRect.width - offset);
    setArrowPosition({
      x: x
    });
  }, [arrow]);
  var placeArrowY = React.useCallback(function (menuY, anchorRect, containerRect, menuRect) {
    if (!arrow) return;
    var y = anchorRect.top - containerRect.top - menuY + anchorRect.height / 2;
    var offset = arrowRef.current.offsetHeight * 1.25;
    y = Math.max(offset, y);
    y = Math.min(y, menuRect.height - offset);
    setArrowPosition({
      y: y
    });
  }, [arrow]);
  var placeLeftorRight = React.useCallback(function (_ref4, _ref5, _ref6) {
    var anchorRect = _ref4.anchorRect,
        containerRect = _ref4.containerRect,
        menuRect = _ref4.menuRect;
    var placeLeftorRightY = _ref5.placeLeftorRightY,
        placeLeftX = _ref5.placeLeftX,
        placeRightX = _ref5.placeRightX;
    var getLeftOverflow = _ref6.getLeftOverflow,
        getRightOverflow = _ref6.getRightOverflow,
        confineHorizontally = _ref6.confineHorizontally,
        confineVertically = _ref6.confineVertically;
    var computedDirection = direction;
    var y = placeLeftorRightY;

    if (position !== 'initial') {
      y = confineVertically(y);

      if (position === 'anchor') {
        y = Math.min(y, anchorRect.bottom - containerRect.top);
        y = Math.max(y, anchorRect.top - containerRect.top - menuRect.height);
      }
    }

    var x, leftOverflow, rightOverflow;

    if (computedDirection === 'left') {
      x = placeLeftX;

      if (position !== 'initial') {
        leftOverflow = getLeftOverflow(x);

        if (leftOverflow < 0) {
          rightOverflow = getRightOverflow(placeRightX);

          if (rightOverflow <= 0 || -leftOverflow > rightOverflow) {
            x = placeRightX;
            computedDirection = 'right';
          }
        }
      }
    } else {
      x = placeRightX;

      if (position !== 'initial') {
        rightOverflow = getRightOverflow(x);

        if (rightOverflow > 0) {
          leftOverflow = getLeftOverflow(placeLeftX);

          if (leftOverflow >= 0 || -leftOverflow < rightOverflow) {
            x = placeLeftX;
            computedDirection = 'left';
          }
        }
      }
    }

    if (position === 'auto') x = confineHorizontally(x);
    placeArrowY(y, anchorRect, containerRect, menuRect);
    return {
      x: x,
      y: y,
      computedDirection: computedDirection
    };
  }, [placeArrowY, direction, position]);
  var placeToporBottom = React.useCallback(function (_ref7, _ref8, _ref9) {
    var anchorRect = _ref7.anchorRect,
        containerRect = _ref7.containerRect,
        menuRect = _ref7.menuRect;
    var placeToporBottomX = _ref8.placeToporBottomX,
        placeTopY = _ref8.placeTopY,
        placeBottomY = _ref8.placeBottomY;
    var getTopOverflow = _ref9.getTopOverflow,
        getBottomOverflow = _ref9.getBottomOverflow,
        confineHorizontally = _ref9.confineHorizontally,
        confineVertically = _ref9.confineVertically;
    var computedDirection = direction === 'top' ? 'top' : 'bottom';
    var x = placeToporBottomX;

    if (position !== 'initial') {
      x = confineHorizontally(x);

      if (position === 'anchor') {
        x = Math.min(x, anchorRect.right - containerRect.left);
        x = Math.max(x, anchorRect.left - containerRect.left - menuRect.width);
      }
    }

    var y, topOverflow, bottomOverflow;

    if (computedDirection === 'top') {
      y = placeTopY;

      if (position !== 'initial') {
        topOverflow = getTopOverflow(y);

        if (topOverflow < 0) {
          bottomOverflow = getBottomOverflow(placeBottomY);

          if (bottomOverflow <= 0 || -topOverflow > bottomOverflow) {
            y = placeBottomY;
            computedDirection = 'bottom';
          }
        }
      }
    } else {
      y = placeBottomY;

      if (position !== 'initial') {
        bottomOverflow = getBottomOverflow(y);

        if (bottomOverflow > 0) {
          topOverflow = getTopOverflow(placeTopY);

          if (topOverflow >= 0 || -topOverflow < bottomOverflow) {
            y = placeTopY;
            computedDirection = 'top';
          }
        }
      }
    }

    if (position === 'auto') y = confineVertically(y);
    placeArrowX(x, anchorRect, containerRect, menuRect);
    return {
      x: x,
      y: y,
      computedDirection: computedDirection
    };
  }, [placeArrowX, direction, position]);
  var positionMenu = React.useCallback(function (positionHelpers, anchorRef) {
    var menuRect = positionHelpers.menuRect,
        containerRect = positionHelpers.containerRect,
        helpers = _objectWithoutPropertiesLoose(positionHelpers, ["menuRect", "containerRect"]);

    var horizontalOffset = offsetX;
    var verticalOffset = offsetY;

    if (arrow) {
      if (direction === 'left' || direction === 'right') {
        horizontalOffset += arrowRef.current.offsetWidth;
      } else {
        verticalOffset += arrowRef.current.offsetHeight;
      }
    }

    var anchorRect = anchorRef.current.getBoundingClientRect();
    var placeLeftX = anchorRect.left - containerRect.left - menuRect.width - horizontalOffset;
    var placeRightX = anchorRect.right - containerRect.left + horizontalOffset;
    var placeTopY = anchorRect.top - containerRect.top - menuRect.height - verticalOffset;
    var placeBottomY = anchorRect.bottom - containerRect.top + verticalOffset;
    var placeToporBottomX, placeLeftorRightY;

    if (align === 'end') {
      placeToporBottomX = anchorRect.right - containerRect.left - menuRect.width;
      placeLeftorRightY = anchorRect.bottom - containerRect.top - menuRect.height;
    } else if (align === 'center') {
      placeToporBottomX = anchorRect.left - containerRect.left - (menuRect.width - anchorRect.width) / 2;
      placeLeftorRightY = anchorRect.top - containerRect.top - (menuRect.height - anchorRect.height) / 2;
    } else {
      placeToporBottomX = anchorRect.left - containerRect.left;
      placeLeftorRightY = anchorRect.top - containerRect.top;
    }

    placeToporBottomX += horizontalOffset;
    placeLeftorRightY += verticalOffset;
    var rects = {
      anchorRect: anchorRect,
      containerRect: containerRect,
      menuRect: menuRect
    };
    var placements = {
      placeLeftX: placeLeftX,
      placeRightX: placeRightX,
      placeLeftorRightY: placeLeftorRightY,
      placeTopY: placeTopY,
      placeBottomY: placeBottomY,
      placeToporBottomX: placeToporBottomX
    };

    switch (direction) {
      case 'left':
      case 'right':
        return placeLeftorRight(rects, placements, helpers);

      case 'top':
      case 'bottom':
      default:
        return placeToporBottom(rects, placements, helpers);
    }
  }, [arrow, align, direction, offsetX, offsetY, placeLeftorRight, placeToporBottom]);
  var positionContextMenu = React.useCallback(function (positionHelpers, anchorPoint) {
    var menuRect = positionHelpers.menuRect,
        containerRect = positionHelpers.containerRect,
        getLeftOverflow = positionHelpers.getLeftOverflow,
        getRightOverflow = positionHelpers.getRightOverflow,
        getTopOverflow = positionHelpers.getTopOverflow,
        getBottomOverflow = positionHelpers.getBottomOverflow,
        confineHorizontally = positionHelpers.confineHorizontally,
        confineVertically = positionHelpers.confineVertically;
    var x, y;
    x = anchorPoint.x - containerRect.left;
    y = anchorPoint.y - containerRect.top;
    var rightOverflow = getRightOverflow(x);

    if (rightOverflow > 0) {
      var adjustedX = x - menuRect.width;
      var leftOverflow = getLeftOverflow(adjustedX);

      if (leftOverflow >= 0 || -leftOverflow < rightOverflow) {
        x = adjustedX;
      }

      x = confineHorizontally(x);
    }

    var computedDirection = 'bottom';
    var bottomOverflow = getBottomOverflow(y);

    if (bottomOverflow > 0) {
      var adjustedY = y - menuRect.height;
      var topOverflow = getTopOverflow(adjustedY);

      if (topOverflow >= 0 || -topOverflow < bottomOverflow) {
        y = adjustedY;
        computedDirection = 'top';
      }

      y = confineVertically(y);
    }

    return {
      x: x,
      y: y,
      computedDirection: computedDirection
    };
  }, []);
  var handlePosition = React.useCallback(function () {
    if (!menuRef.current) {
      !isProd && console.warn('Menu ref is null and might not be positioned properly. You could report an issue on GitHub.');
      return;
    }

    if (!containerRef.current) {
      !isProd && console.error('Menu cannot be positioned properly as container ref is null.', 'If you initialise isOpen prop to true for ControlledMenu, please see this link for a solution: https://github.com/szhsin/react-menu/issues/2#issuecomment-719166062');
      return;
    }

    var helpers = positionHelpers(boundingBoxRef, boundingBoxPadding);
    var menuRect = helpers.menuRect;
    var results = {
      computedDirection: 'bottom'
    };

    if (anchorPoint) {
      results = positionContextMenu(helpers, anchorPoint);
    } else if (anchorRef) {
      results = positionMenu(helpers, anchorRef);
    }

    var _results = results,
        x = _results.x,
        y = _results.y,
        computedDirection = _results.computedDirection;

    if (overflow !== 'visible') {
      var getTopOverflow = helpers.getTopOverflow,
          getBottomOverflow = helpers.getBottomOverflow;
      setMaxHeight(function (height) {
        var newHeight = -1;
        var bottomOverflow = getBottomOverflow(y);

        if (bottomOverflow > 0 || floatEqual(bottomOverflow, 0) && height >= 0) {
          newHeight = menuRect.height - bottomOverflow;
        } else {
          var topOverflow = getTopOverflow(y);

          if (topOverflow < 0 || floatEqual(topOverflow, 0) && height >= 0) {
            newHeight = menuRect.height + topOverflow;
            if (newHeight >= 0) y -= topOverflow;
          }
        }

        return newHeight;
      });
    }

    setMenuPosition({
      x: x,
      y: y
    });
    setExpandedDirection(computedDirection);
    latestMenuSize.current = {
      width: menuRect.width,
      height: menuRect.height
    };
  }, [anchorPoint, anchorRef, containerRef, boundingBoxRef, boundingBoxPadding, overflow, positionHelpers, positionMenu, positionContextMenu]);
  useIsomorphicLayoutEffect(function () {
    if (isOpen) {
      handlePosition();
      if (latestOpen.current) forceReposSubmenu();
    }

    latestHandlePosition.current = handlePosition;
  }, [isOpen, handlePosition, reposFlag]);
  useIsomorphicLayoutEffect(function () {
    if (animation) {
      if (isOpen) {
        setClosing(false);
      } else if (isOpen !== latestOpen.current) {
        setClosing(true);
      }
    }

    latestOpen.current = isOpen;
  }, [animation, isOpen]);
  React.useEffect(function () {
    if (!isOpen || viewScroll === 'initial') return;
    var scroll = viewScroll;
    if (scroll === 'auto' && overflow !== 'visible') scroll = 'close';

    var handleScroll = function handleScroll() {
      if (scroll === 'auto') {
        handlePosition();
      } else {
        safeCall(onClose, {
          reason: CloseReason.SCROLL
        });
      }
    };

    var target = scrollingRef.current.addEventListener ? scrollingRef.current : window;
    target.addEventListener('scroll', handleScroll);
    return function () {
      return target.removeEventListener('scroll', handleScroll);
    };
  }, [scrollingRef, isOpen, overflow, onClose, viewScroll, handlePosition]);
  React.useEffect(function () {
    if (typeof ResizeObserver !== 'function' || reposition === 'initial') return;
    var resizeObserver = new ResizeObserver(function (_ref10) {
      var entry = _ref10[0];
      var borderBoxSize = entry.borderBoxSize,
          target = entry.target;
      var width, height;

      if (borderBoxSize) {
        var _ref11 = borderBoxSize[0] || borderBoxSize,
            inlineSize = _ref11.inlineSize,
            blockSize = _ref11.blockSize;

        width = inlineSize;
        height = blockSize;
      } else {
        var borderRect = target.getBoundingClientRect();
        width = borderRect.width;
        height = borderRect.height;
      }

      if (width === 0 || height === 0) return;
      if (floatEqual(width, latestMenuSize.current.width) && floatEqual(height, latestMenuSize.current.height)) return;
      latestHandlePosition.current();
      forceReposSubmenu();
    });
    var observeTarget = menuRef.current;
    resizeObserver.observe(observeTarget, {
      box: 'border-box'
    });
    return function () {
      return resizeObserver.unobserve(observeTarget);
    };
  }, [reposition]);
  React.useEffect(function () {
    if (!isOpen) {
      dispatch({
        type: HoverIndexActionTypes.RESET
      });
      if (!animation) setMaxHeight(-1);
    }

    var id = setTimeout(function () {
      if (!isOpen || !menuRef.current || menuRef.current.contains(document.activeElement)) return;
      if (captureFocus) menuRef.current.focus();

      if (menuItemFocus.position === FocusPositions.FIRST) {
        dispatch({
          type: HoverIndexActionTypes.FIRST
        });
      } else if (menuItemFocus.position === FocusPositions.LAST) {
        dispatch({
          type: HoverIndexActionTypes.LAST
        });
      }
    }, animation ? 150 : 100);
    return function () {
      return clearTimeout(id);
    };
  }, [animation, captureFocus, isOpen, menuItemFocus]);
  var isSubmenuOpen = openSubmenuCount > 0;
  var context = React.useMemo(function () {
    return {
      isParentOpen: isOpen,
      hoverIndex: hoverIndex,
      isSubmenuOpen: isSubmenuOpen,
      dispatch: dispatch
    };
  }, [isOpen, hoverIndex, isSubmenuOpen]);
  var modifiers = {
    open: isOpen,
    closing: isClosing,
    animation: animation,
    dir: animation && expandedDirection
  };
  var userModifiers = Object.freeze(_extends({}, modifiers, {
    dir: expandedDirection
  }));
  var arrowModifiers = Object.freeze({
    dir: expandedDirection
  });
  var overflowStyles = null;

  if (maxHeight >= 0) {
    overflowStyles = {
      maxHeight: maxHeight,
      overflow: overflow
    };
  }

  var handlers = attachHandlerProps({
    onKeyDown: handleKeyDown,
    onAnimationEnd: handleAnimationEnd
  }, restProps);
  return /*#__PURE__*/React__default.createElement("ul", _extends({
    role: "menu",
    tabIndex: "-1",
    "aria-disabled": isDisabled || undefined,
    "aria-label": ariaLabel
  }, restProps, handlers, {
    ref: menuRef,
    className: bem(menuClass, null, modifiers)(className, userModifiers),
    style: _extends({}, flatStyles(styles, userModifiers), overflowStyles, {
      left: menuPosition.x + "px",
      top: menuPosition.y + "px"
    })
  }), arrow && /*#__PURE__*/React__default.createElement("div", {
    className: bem(menuClass, menuArrowClass, arrowModifiers)(arrowClassName),
    style: _extends({}, flatStyles(arrowStyles, arrowModifiers), {
      left: arrowPosition.x && arrowPosition.x + "px",
      top: arrowPosition.y && arrowPosition.y + "px"
    }),
    ref: arrowRef,
    role: "presentation"
  }), /*#__PURE__*/React__default.createElement(MenuListContext.Provider, {
    value: reposSubmenu
  }, /*#__PURE__*/React__default.createElement(MenuListItemContext.Provider, {
    value: context
  }, menuItems)));
}), 'MenuList');

function submenuCountReducer(state, _ref12) {
  var type = _ref12.type;

  switch (type) {
    case SubmenuActionTypes.OPEN:
      return state + 1;

    case SubmenuActionTypes.CLOSE:
      return Math.max(state - 1, 0);

    default:
      return state;
  }
}

var useMenuList = function useMenuList(menuListProps, _ref) {
  var id = _ref.id,
      animation = _ref.animation,
      boundingBoxRef = _ref.boundingBoxRef,
      boundingBoxPadding = _ref.boundingBoxPadding,
      debugging = _ref.debugging,
      reposition = _ref.reposition,
      viewScroll = _ref.viewScroll,
      portal = _ref.portal,
      theming = _ref.theming,
      isMounted = _ref.isMounted,
      onClick = _ref.onClick,
      onClose = _ref.onClose,
      skipClick = _ref.skipClick;
  var containerRef = React.useRef(null);
  var scrollingRef = React.useRef(null);
  var settings = React.useMemo(function () {
    return {
      animation: animation,
      boundingBoxRef: boundingBoxRef,
      boundingBoxPadding: boundingBoxPadding,
      menuRootRef: containerRef,
      scrollingRef: scrollingRef,
      debugging: debugging,
      reposition: reposition,
      viewScroll: viewScroll
    };
  }, [animation, boundingBoxRef, boundingBoxPadding, debugging, reposition, viewScroll]);
  var eventHandlers = React.useMemo(function () {
    return {
      handleClick: function handleClick(event, isStopPropagation, isCheckorRadio) {
        if (!isStopPropagation) safeCall(onClick, event);
        var keepOpen = event.keepOpen;

        if (keepOpen === undefined) {
          keepOpen = isCheckorRadio && event.key === Keys.SPACE;
        }

        if (!keepOpen) {
          safeCall(onClose, {
            value: event.value,
            key: event.key,
            reason: CloseReason.CLICK
          });
        }
      },
      handleClose: function handleClose(key) {
        safeCall(onClose, {
          key: key,
          reason: CloseReason.CLICK
        });
      }
    };
  }, [onClick, onClose]);

  var handleKeyDown = function handleKeyDown(_ref2) {
    var key = _ref2.key;

    switch (key) {
      case Keys.ESC:
        safeCall(onClose, {
          key: key,
          reason: CloseReason.CANCEL
        });
        break;
    }
  };

  var handleBlur = function handleBlur(e) {
    if (menuListProps.isOpen && !e.currentTarget.contains(e.relatedTarget || document.activeElement) && !debugging) {
      safeCall(onClose, {
        reason: CloseReason.BLUR
      });

      if (skipClick) {
        skipClick.current = true;
        setTimeout(function () {
          return skipClick.current = false;
        }, 300);
      }
    }
  };

  var menuList = /*#__PURE__*/React__default.createElement("div", {
    id: id,
    className: bem(menuContainerClass, null, {
      theme: theming
    })(),
    role: "presentation",
    ref: containerRef,
    onKeyDown: handleKeyDown,
    onBlur: handleBlur
  }, isMounted && /*#__PURE__*/React__default.createElement(SettingsContext.Provider, {
    value: settings
  }, /*#__PURE__*/React__default.createElement(EventHandlersContext.Provider, {
    value: eventHandlers
  }, /*#__PURE__*/React__default.createElement(MenuList, _extends({}, menuListProps, {
    containerRef: containerRef,
    onClose: onClose
  })))));

  if (portal) {
    return /*#__PURE__*/ReactDOM.createPortal(menuList, document.body);
  } else {
    return menuList;
  }
};

var Menu = /*#__PURE__*/React__default.memo(function Menu(_ref) {
  var ariaLabel = _ref['aria-label'],
      id = _ref.id,
      animation = _ref.animation,
      boundingBoxRef = _ref.boundingBoxRef,
      boundingBoxPadding = _ref.boundingBoxPadding,
      debugging = _ref.debugging,
      reposition = _ref.reposition,
      viewScroll = _ref.viewScroll,
      keepMounted = _ref.keepMounted,
      menuButton = _ref.menuButton,
      portal = _ref.portal,
      theming = _ref.theming,
      onClick = _ref.onClick,
      onChange = _ref.onChange,
      restProps = _objectWithoutPropertiesLoose(_ref, ["aria-label", "id", "animation", "boundingBoxRef", "boundingBoxPadding", "captureFocus", "debugging", "reposition", "viewScroll", "keepMounted", "menuButton", "portal", "theming", "onClick", "onChange"]);

  var _useMenuState = useMenuState(keepMounted),
      isMounted = _useMenuState.isMounted,
      isOpen = _useMenuState.isOpen,
      menuItemFocus = _useMenuState.menuItemFocus,
      openMenu = _useMenuState.openMenu,
      closeMenu = _useMenuState.closeMenu;

  var skipClick = React.useRef(false);
  var buttonRef = React.useRef(null);
  var button = safeCall(menuButton, {
    open: isOpen
  });
  if (!button) throw new Error('Menu requires a menuButton prop.');
  var handleClose = React.useCallback(function (e) {
    closeMenu();
    if (e.key) buttonRef.current.focus();
  }, [closeMenu]);
  var handleClick = React.useCallback(function (e) {
    if (skipClick.current) return;
    openMenu(e.detail === 0 ? FocusPositions.FIRST : FocusPositions.INITIAL);
  }, [openMenu]);
  var handleKeyDown = React.useCallback(function (e) {
    var handled = false;

    switch (e.key) {
      case Keys.UP:
        openMenu(FocusPositions.LAST);
        handled = true;
        break;

      case Keys.DOWN:
        openMenu(FocusPositions.FIRST);
        handled = true;
        break;
    }

    if (handled) e.preventDefault();
  }, [openMenu]);
  var renderButton = React.useMemo(function () {
    var buttonProps = _extends({
      ref: buttonRef
    }, attachHandlerProps({
      onClick: handleClick,
      onKeyDown: handleKeyDown
    }, button.props));

    if (getName(button.type) === 'MenuButton') {
      buttonProps.isOpen = isOpen;
    }

    return /*#__PURE__*/React__default.cloneElement(button, buttonProps);
  }, [button, isOpen, handleClick, handleKeyDown]);
  var menuList = useMenuList(_extends({}, restProps, {
    ariaLabel: ariaLabel || (typeof button.props.children === 'string' ? button.props.children : 'Menu'),
    anchorRef: buttonRef,
    isOpen: isOpen,
    menuItemFocus: menuItemFocus
  }), {
    id: id,
    animation: animation,
    boundingBoxRef: boundingBoxRef,
    boundingBoxPadding: boundingBoxPadding,
    debugging: debugging,
    reposition: reposition,
    viewScroll: viewScroll,
    portal: portal,
    theming: theming,
    isMounted: isMounted,
    onClick: onClick,
    onClose: handleClose,
    skipClick: skipClick
  });
  useMenuChange(onChange, isOpen);
  return /*#__PURE__*/React__default.createElement(React__default.Fragment, null, renderButton, menuList);
});
Menu.propTypes = _extends({}, menuPropTypesBase, {
  keepMounted: PropTypes.bool,
  menuButton: PropTypes.oneOfType([PropTypes.element, PropTypes.func]).isRequired,
  onChange: PropTypes.func
});
Menu.defaultProps = _extends({}, menuDefaultPropsBase, {
  keepMounted: true
});

var ControlledMenu = /*#__PURE__*/React__default.memo(function ControlledMenu(_ref) {
  var ariaLabel = _ref['aria-label'],
      id = _ref.id,
      animation = _ref.animation,
      boundingBoxRef = _ref.boundingBoxRef,
      boundingBoxPadding = _ref.boundingBoxPadding,
      debugging = _ref.debugging,
      reposition = _ref.reposition,
      viewScroll = _ref.viewScroll,
      portal = _ref.portal,
      theming = _ref.theming,
      isMounted = _ref.isMounted,
      onClick = _ref.onClick,
      onClose = _ref.onClose,
      restProps = _objectWithoutPropertiesLoose(_ref, ["aria-label", "id", "animation", "boundingBoxRef", "boundingBoxPadding", "debugging", "reposition", "viewScroll", "portal", "theming", "isMounted", "onClick", "onClose"]);

  return useMenuList(_extends({}, restProps, {
    ariaLabel: ariaLabel || 'Menu'
  }), {
    id: id,
    animation: animation,
    boundingBoxRef: boundingBoxRef,
    boundingBoxPadding: boundingBoxPadding,
    debugging: debugging,
    reposition: reposition,
    viewScroll: viewScroll,
    portal: portal,
    theming: theming,
    isMounted: isMounted,
    onClick: onClick,
    onClose: onClose
  });
});
ControlledMenu.propTypes = _extends({}, menuPropTypesBase, {
  anchorPoint: PropTypes.exact({
    x: PropTypes.number,
    y: PropTypes.number
  }),
  anchorRef: PropTypes.object,
  captureFocus: PropTypes.bool,
  isOpen: PropTypes.bool,
  isMounted: PropTypes.bool,
  menuItemFocus: PropTypes.exact({
    position: PropTypes.string
  }),
  onClose: PropTypes.func
});
ControlledMenu.defaultProps = _extends({}, menuDefaultPropsBase, {
  captureFocus: true,
  isOpen: false,
  isMounted: true,
  menuItemFocus: {
    position: FocusPositions.INITIAL
  }
});

var SubMenu = defineName( /*#__PURE__*/React__default.memo(function SubMenu(_ref) {
  var ariaLabel = _ref['aria-label'],
      itemClassName = _ref.itemClassName,
      itemStyles = _ref.itemStyles,
      disabled = _ref.disabled,
      keepMounted = _ref.keepMounted,
      label = _ref.label,
      index = _ref.index,
      onChange = _ref.onChange,
      restProps = _objectWithoutPropertiesLoose(_ref, ["aria-label", "itemClassName", "itemStyles", "disabled", "keepMounted", "label", "index", "onChange", "captureFocus", "repositionFlag"]);

  var _useMenuState = useMenuState(keepMounted),
      isMounted = _useMenuState.isMounted,
      isOpen = _useMenuState.isOpen,
      menuItemFocus = _useMenuState.menuItemFocus,
      openMenu = _useMenuState.openMenu,
      closeMenu = _useMenuState.closeMenu;

  var _useContext = React.useContext(MenuListItemContext),
      isParentOpen = _useContext.isParentOpen,
      hoverIndex = _useContext.hoverIndex,
      isSubmenuOpen = _useContext.isSubmenuOpen,
      dispatch = _useContext.dispatch;

  var _useContext2 = React.useContext(SettingsContext),
      debugging = _useContext2.debugging;

  var isHovering = hoverIndex === index;
  var isDisabled = Boolean(disabled);

  var _useActiveState = useActiveState(isHovering, isDisabled, Keys.RIGHT),
      isActive = _useActiveState.isActive,
      onKeyUp = _useActiveState.onKeyUp,
      activeStateHandlers = _objectWithoutPropertiesLoose(_useActiveState, ["isActive", "onKeyUp"]);

  var containerRef = React.useRef(null);
  var itemRef = React.useRef(null);
  var timeoutId = React.useRef();
  var handleClose = React.useCallback(function () {
    itemRef.current.focus();
  }, []);

  var handleMouseEnter = function handleMouseEnter() {
    if (isDisabled || isOpen) return;
    if (!isSubmenuOpen) dispatch({
      type: HoverIndexActionTypes.SET,
      index: index
    });
    timeoutId.current = setTimeout(function () {
      dispatch({
        type: HoverIndexActionTypes.SET,
        index: index
      });
      timeoutId.current = setTimeout(openMenu, SUBMENU_OPEN_DELAY);
    }, SUBMENU_CLOSE_DELAY);
  };

  var handleMouseLeave = function handleMouseLeave() {
    clearTimeout(timeoutId.current);

    if (!isOpen) {
      dispatch({
        type: HoverIndexActionTypes.UNSET,
        index: index
      });
    }
  };

  var handleKeyDown = function handleKeyDown(e) {
    var handled = false;

    switch (e.key) {
      case Keys.LEFT:
        if (isOpen) {
          handleClose();
          handled = true;
        }

        break;

      case Keys.RIGHT:
        if (!isOpen) handled = true;
        break;
    }

    if (handled) {
      e.preventDefault();
      e.stopPropagation();
    }
  };

  var handleKeyUp = function handleKeyUp(e) {
    if (!isActive) return;
    onKeyUp(e);

    switch (e.key) {
      case Keys.SPACE:
      case Keys.ENTER:
      case Keys.RIGHT:
        openMenu(FocusPositions.FIRST);
        break;
    }
  };

  var handleBlur = function handleBlur(e) {
    if (debugging) return;
    var relatedTarget = e.relatedTarget || document.activeElement;

    if (!e.currentTarget.contains(relatedTarget)) {
      closeMenu();
      dispatch({
        type: HoverIndexActionTypes.UNSET,
        index: index
      });
    } else if (itemRef.current.contains(relatedTarget)) {
      closeMenu();
    }
  };

  React.useEffect(function () {
    if (isHovering && isParentOpen) {
      itemRef.current.focus();
    } else {
      closeMenu();
    }
  }, [isHovering, isParentOpen, closeMenu]);
  React.useEffect(function () {
    dispatch({
      type: isOpen ? SubmenuActionTypes.OPEN : SubmenuActionTypes.CLOSE
    });
  }, [dispatch, isOpen]);
  useMenuChange(onChange, isOpen);
  var modifiers = Object.freeze({
    open: isOpen,
    hover: isHovering,
    active: isActive,
    disabled: isDisabled
  });
  return /*#__PURE__*/React__default.createElement("li", {
    className: bem(menuClass, subMenuClass)(),
    role: "presentation",
    ref: containerRef,
    onKeyDown: handleKeyDown,
    onBlur: handleBlur
  }, /*#__PURE__*/React__default.createElement("div", _extends({
    className: bem(menuClass, menuItemClass, modifiers)(itemClassName),
    style: flatStyles(itemStyles, modifiers),
    role: "menuitem",
    "aria-haspopup": "true",
    "aria-expanded": isOpen,
    "aria-disabled": isDisabled || undefined,
    tabIndex: isHovering && !isOpen ? 0 : -1,
    ref: itemRef,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onClick: function onClick() {
      return !isDisabled && openMenu();
    },
    onKeyUp: handleKeyUp
  }, activeStateHandlers), safeCall(label, modifiers)), isMounted && /*#__PURE__*/React__default.createElement(MenuList, _extends({}, restProps, {
    ariaLabel: ariaLabel || (typeof label === 'string' ? label : 'Submenu'),
    anchorRef: itemRef,
    containerRef: containerRef,
    isOpen: isOpen,
    isDisabled: isDisabled,
    menuItemFocus: menuItemFocus,
    onClose: handleClose
  })));
}), 'SubMenu');
SubMenu.propTypes = _extends({}, sharedMenuPropTypes, stylePropTypes('item'), {
  disabled: PropTypes.bool,
  keepMounted: PropTypes.bool,
  label: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  onChange: PropTypes.func
});
SubMenu.defaultProps = _extends({}, sharedMenuDefaultProp, {
  direction: 'right',
  keepMounted: true
});

var MenuItem = defineName( /*#__PURE__*/React__default.memo(function MenuItem(_ref) {
  var className = _ref.className,
      styles = _ref.styles,
      value = _ref.value,
      href = _ref.href,
      type = _ref.type,
      checked = _ref.checked,
      disabled = _ref.disabled,
      index = _ref.index,
      children = _ref.children,
      onClick = _ref.onClick,
      restProps = _objectWithoutPropertiesLoose(_ref, ["className", "styles", "value", "href", "type", "checked", "disabled", "index", "children", "onClick"]);

  var isDisabled = Boolean(disabled);

  var _useItemState = useItemState(isDisabled, index),
      ref = _useItemState.ref,
      isHovering = _useItemState.isHovering,
      onBlur = _useItemState.onBlur,
      onMouseEnter = _useItemState.onMouseEnter,
      onMouseLeave = _useItemState.onMouseLeave;

  var eventHandlers = React.useContext(EventHandlersContext);
  var radioGroup = React.useContext(RadioGroupContext);

  var _useActiveState = useActiveState(isHovering, isDisabled),
      isActive = _useActiveState.isActive,
      onKeyUp = _useActiveState.onKeyUp,
      activeStateBlur = _useActiveState.onBlur,
      activeStateHandlers = _objectWithoutPropertiesLoose(_useActiveState, ["isActive", "onKeyUp", "onBlur"]);

  var isRadio = type === 'radio';
  var isCheckBox = type === 'checkbox';
  var isAnchor = Boolean(href) && !isDisabled && !isRadio && !isCheckBox;
  var isChecked = isRadio ? radioGroup.value === value : isCheckBox ? Boolean(checked) : false;

  var handleClick = function handleClick(key) {
    if (isDisabled) return;
    var isStopPropagation = false;
    var event = {
      value: value,
      key: key
    };

    if (isRadio) {
      event.name = radioGroup.name;
      safeCall(radioGroup.onChange, event);
    }

    event.checked = isCheckBox ? !isChecked : false;
    isStopPropagation = safeCall(onClick, event) === false;
    eventHandlers.handleClick(event, isStopPropagation, isCheckBox || isRadio);
  };

  var handleKeyUp = function handleKeyUp(e) {
    if (!isActive) return;
    onKeyUp(e);

    switch (e.key) {
      case Keys.SPACE:
      case Keys.ENTER:
        if (isAnchor) {
          ref.current.click();
        } else {
          handleClick(e.key);
        }

        break;
    }
  };

  var handleBlur = function handleBlur(e) {
    activeStateBlur(e);
    onBlur(e);
  };

  var modifiers = Object.freeze({
    type: type,
    disabled: isDisabled,
    hover: isHovering,
    active: isActive,
    checked: isChecked,
    anchor: isAnchor
  });
  var handlers = attachHandlerProps(_extends({}, activeStateHandlers, {
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onKeyUp: handleKeyUp,
    onBlur: handleBlur,
    onClick: function onClick() {
      return handleClick();
    }
  }), restProps);

  var menuItemProps = _extends({
    role: isRadio ? 'menuitemradio' : isCheckBox ? 'menuitemcheckbox' : 'menuitem',
    'aria-checked': isRadio || isCheckBox ? isChecked : undefined,
    'aria-disabled': isDisabled || undefined,
    tabIndex: isHovering ? 0 : -1
  }, restProps, handlers, {
    ref: ref,
    className: bem(menuClass, menuItemClass, modifiers)(className),
    style: flatStyles(styles, modifiers)
  });

  var renderChildren = safeCall(children, modifiers);

  if (isAnchor) {
    return /*#__PURE__*/React__default.createElement("li", {
      role: "presentation"
    }, /*#__PURE__*/React__default.createElement("a", _extends({}, menuItemProps, {
      href: href
    }), renderChildren));
  } else {
    return /*#__PURE__*/React__default.createElement("li", menuItemProps, renderChildren);
  }
}), 'MenuItem');
MenuItem.propTypes = _extends({}, stylePropTypes(), {
  value: PropTypes.any,
  href: PropTypes.string,
  type: PropTypes.oneOf(['checkbox', 'radio']),
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  onClick: PropTypes.func
});

var FocusableItem = defineName( /*#__PURE__*/React__default.memo(function FocusableItem(_ref) {
  var className = _ref.className,
      styles = _ref.styles,
      disabled = _ref.disabled,
      index = _ref.index,
      children = _ref.children,
      restProps = _objectWithoutPropertiesLoose(_ref, ["className", "styles", "disabled", "index", "children"]);

  var isDisabled = Boolean(disabled);

  var _useItemState = useItemState(isDisabled, index),
      ref = _useItemState.ref,
      isHovering = _useItemState.isHovering,
      setHover = _useItemState.setHover,
      onBlur = _useItemState.onBlur,
      onMouseEnter = _useItemState.onMouseEnter,
      _onMouseLeave = _useItemState.onMouseLeave;

  var _useContext = React.useContext(EventHandlersContext),
      handleClose = _useContext.handleClose;

  var baseParams = {
    disabled: isDisabled,
    hover: isHovering
  };
  var modifiers = Object.freeze(_extends({}, baseParams, {
    focusable: true
  }));
  var renderChildren = safeCall(children, _extends({}, baseParams, {
    ref: ref,
    closeMenu: handleClose
  }));
  var handlers = attachHandlerProps({
    onMouseEnter: onMouseEnter,
    onMouseLeave: function onMouseLeave(e) {
      return _onMouseLeave(e, true);
    },
    onFocus: setHover,
    onBlur: onBlur
  }, restProps);
  return /*#__PURE__*/React__default.createElement("li", _extends({
    "aria-disabled": isDisabled || undefined,
    role: "menuitem",
    tabIndex: "-1"
  }, restProps, handlers, {
    className: bem(menuClass, menuItemClass, modifiers)(className),
    style: flatStyles(styles, modifiers)
  }), renderChildren);
}), 'FocusableItem');
FocusableItem.propTypes = _extends({}, stylePropTypes(), {
  disabled: PropTypes.bool,
  children: PropTypes.func.isRequired
});

var MenuDivider = defineName( /*#__PURE__*/React__default.memo(function MenuDivider(_ref) {
  var className = _ref.className,
      styles = _ref.styles,
      restProps = _objectWithoutPropertiesLoose(_ref, ["className", "styles"]);

  return /*#__PURE__*/React__default.createElement("li", _extends({
    role: "separator"
  }, restProps, {
    className: bem(menuClass, menuDividerClass)(className),
    style: flatStyles(styles)
  }));
}), 'MenuDivider');
MenuDivider.propTypes = _extends({}, stylePropTypes());

var MenuHeader = defineName( /*#__PURE__*/React__default.memo(function MenuHeader(_ref) {
  var className = _ref.className,
      styles = _ref.styles,
      children = _ref.children,
      restProps = _objectWithoutPropertiesLoose(_ref, ["className", "styles", "children"]);

  return /*#__PURE__*/React__default.createElement("li", _extends({
    role: "presentation"
  }, restProps, {
    className: bem(menuClass, menuHeaderClass)(className),
    style: flatStyles(styles)
  }), children);
}), 'MenuHeader');
MenuHeader.propTypes = _extends({}, stylePropTypes());

var MenuRadioGroup = defineName( /*#__PURE__*/React__default.memo(function MenuRadioGroup(_ref) {
  var ariaLabel = _ref['aria-label'],
      className = _ref.className,
      styles = _ref.styles,
      name = _ref.name,
      value = _ref.value,
      children = _ref.children,
      onChange = _ref.onChange,
      restProps = _objectWithoutPropertiesLoose(_ref, ["aria-label", "className", "styles", "name", "value", "children", "onChange"]);

  var contextValue = React.useMemo(function () {
    return {
      name: name,
      value: value,
      onChange: onChange
    };
  }, [name, value, onChange]);
  return /*#__PURE__*/React__default.createElement("li", {
    role: "presentation"
  }, /*#__PURE__*/React__default.createElement("ul", _extends({
    role: "group",
    "aria-label": ariaLabel || name || 'Radio group'
  }, restProps, {
    className: bem(menuClass, radioGroupClass)(className),
    style: flatStyles(styles)
  }), /*#__PURE__*/React__default.createElement(RadioGroupContext.Provider, {
    value: contextValue
  }, children)));
}), 'MenuRadioGroup');
MenuRadioGroup.propTypes = _extends({}, stylePropTypes(), {
  'aria-label': PropTypes.string,
  name: PropTypes.string,
  value: PropTypes.any,
  children: PropTypes.node.isRequired,
  onChange: PropTypes.func
});

exports.ControlledMenu = ControlledMenu;
exports.FocusableItem = FocusableItem;
exports.Menu = Menu;
exports.MenuButton = MenuButton;
exports.MenuDivider = MenuDivider;
exports.MenuHeader = MenuHeader;
exports.MenuItem = MenuItem;
exports.MenuRadioGroup = MenuRadioGroup;
exports.SubMenu = SubMenu;
exports.applyHOC = applyHOC;
exports.applyStatics = applyStatics;
exports.useMenuState = useMenuState;
